### DISCUSSION TITLE
Custom API key policies, nonauthenticated users, possible?

### DISCUSSION THREAD

**Discussion by flxmrk** [supabase/supabase#10441] [Questions] [answered]
Lets say I build an app, that has authenticated users, and those authenticated users can create various data. Then they have access to a module that lets them generate API keys, then they want those API keys to access (read/update) the data belonging to that user.

Is this kind of scenario simply not possible with supabase?

**Comment by GaryAustin1** [comment, marked as answer]
Probably possible.  RLS is pretty flexible.  You would need some table to associate keys with with the data they can access (possibly mapping to the original user id that generated them. 
 
Then, depending on security needs, you could just pass them in as a custom header and use PostgREST's ability to let you access headers in the RLS to control privileges.   Or you can pass them in a an rpc call as a parameter with your data returned by the function.  This is only safe if the keys are not passed around.  You could certainly expire the keys or delete them from the table after a time.  This is similar to how signed URL's work for storage only you pass them in a header or as a parameter to rpc.

You could also get more sophisticated and use rpc and functions to generate a signed jwt that expires and then use that with a function in an rpc call or rls to verify the jwt is valid.  

PostgREST http context (reading headers) https://postgrest.org/en/stable/api.html#http-context

**Reply by flxmrk** [reply]
ahh very interesting... quite new to postgres and all this way of thinking... didn't know we had access to anything in request beyond what supabase describes.  Thanks for the tip and the docs link! 

So I'd have to write a policy that looked up the provided API KEY, to get the real USER it belongs to, then to give read access to all records with that userid defined (requires the userid also being set per record) - right?

RPC is an interesting route I could see as being useful for some things for sure 🤔

**Reply by GaryAustin1** [reply]
You certainly would need some way to associate the token with the data inorder to use RLS.  That could be user id in the row, or an intermediate foreign key to a table with valid tokens for each row, or some other intermediate identifier.    It all depends on if you have data that the real user should be limited to versus access with the token.  If your "shared" data is all in separate tables from "real" user data then life would be easier.

**Comment by flxmrk** [comment]
@GaryAustin1  I defined my key in the createClient options.global.headers object - and it works with standard rest calls via the supabase js client, but the headers do not seem to be passed when using .subscribe() for realtime data.  

Do you happen to know if what you described also works for realtime data?  If the websocket (wss://) network call doesn't include the headers, i dont see how a postgres policy will be able to find them.  Even the supabase "apikey" header normally present in rest calls gets moved to the query string on the websocket...

**Reply by flxmrk** [reply]
i do see i could pass them in createClient options.realtime.params object but that seems awful messy?  now i have to write a policy that checks for the header OR a querystring variable?

**Reply by GaryAustin1** [reply]
The same token gets passes to realtime as the authentication header uses.  With auth on a browser that happens automatically.    The RLS policy is just the normal policy.

**Reply by flxmrk** [reply]
Not according to the devtools
![DevTools - 10 0 2 20:3004 2022-12-28 at 12 34 22 PM](https://user-images.githubusercontent.com/2659615/209857628-7659c584-cc0b-43a3-a7d4-4c2a4011f59a.jpg)

**Reply by flxmrk** [reply]
in contrast, with every REST call the headers i added get passed.

not sure how the 'realtime' would have access to headers that were not passed during the `wss://` request?

**Reply by GaryAustin1** [reply]
It is passed as a message, either token, or access_token I don’t remember which.

**Reply by flxmrk** [reply]
hmm, I do see access_token in the payloads, but i've searched quite a few of the messages (including the first message where you might expect it to be) passed over websocket, none of them container my headers

**Reply by GaryAustin1** [reply]
I just reread the thread.  You used the custom header approach which I mentioned works with PostgREST, which is the database API.   In my last responses I was thinking of a custom JWT and thought you were trying to pass that with clientInit.  Realtime will not get any customer header info. It only works with JWT claims which it uses to access the database to test RLS. 

Also I have not personally tested custom claims with the new cloud Realtime and have never seen it’s code, so can only guess it passes thru JWT claims other than role and sub (uuid).
